tests/components/cloudflare/test_config_flow.py

```diff
@@ -0,0 +1,166 @@
+"""Test the Cloudflare config flow."""
+from pycfdns.exceptions import (
+    CloudflareAuthenticationException,
+    CloudflareConnectionException,
+    CloudflareZoneException,
+)
+
+from homeassistant.components.cloudflare.const import CONF_RECORDS, DOMAIN
+from homeassistant.config_entries import SOURCE_USER
+from homeassistant.const import CONF_API_TOKEN, CONF_SOURCE, CONF_ZONE
+from homeassistant.data_entry_flow import (
+    RESULT_TYPE_ABORT,
+    RESULT_TYPE_CREATE_ENTRY,
+    RESULT_TYPE_FORM,
+)
+from homeassistant.setup import async_setup_component
+
+from . import (
+    ENTRY_CONFIG,
+    USER_INPUT,
+    USER_INPUT_RECORDS,
+    USER_INPUT_ZONE,
+    _patch_async_setup,
+    _patch_async_setup_entry,
+)
+
+from tests.common import MockConfigEntry
+
+
+async def test_user_form(hass, cfupdate_flow):
+    """Test we get the user initiated form."""
+    await async_setup_component(hass, "persistent_notification", {})
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={CONF_SOURCE: SOURCE_USER}
+    )
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+    assert result["errors"] == {}
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        USER_INPUT,
+    )
+    await hass.async_block_till_done()
+
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["step_id"] == "zone"
+    assert result["errors"] == {}
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        USER_INPUT_ZONE,
+    )
+    await hass.async_block_till_done()
+
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["step_id"] == "records"
+    assert result["errors"] == {}
+
+    with _patch_async_setup() as mock_setup, _patch_async_setup_entry() as mock_setup_entry:
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            USER_INPUT_RECORDS,
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == RESULT_TYPE_CREATE_ENTRY
+    assert result["title"] == USER_INPUT_ZONE[CONF_ZONE]
+
+    assert result["data"]
+    assert result["data"][CONF_API_TOKEN] == USER_INPUT[CONF_API_TOKEN]
+    assert result["data"][CONF_ZONE] == USER_INPUT_ZONE[CONF_ZONE]
+    assert result["data"][CONF_RECORDS] == USER_INPUT_RECORDS[CONF_RECORDS]
+
+    assert result["result"]
+    assert result["result"].unique_id == USER_INPUT_ZONE[CONF_ZONE]
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6372789)